cinn/frontend/pass/fill_constant_folding.cc

```diff
@@ -0,0 +1,192 @@
+      if (!fill_constant_set.count(key)) {
+        VLOG(4) << "The fill_constant, whose output is Var [" << instr->outputs[0]->id
+                << "], cannot remove because it is the first fill_costant ! ";
+        // retain the first fill constant op node
+        fill_constant_set.emplace(key, &instr->outputs[0]);
+        continue;
+      }
+
+      if (fetch_ids.find(instr->outputs[0]->id) != fetch_ids.end()) {
+        // the fill constant's output variable was fetched, skip
+        VLOG(4) << "Cannot remove fill_constant, because Var [" << instr->outputs[0]->id
+                << "] was fetched by other op ! ";
+        continue;
+      }
+
+      VLOG(4) << "Try remove fill_constant, whose output is Var [" << instr->outputs[0]->id << "]. ";
+      remove_instrs.insert(&instr);
+
+      auto constant_name = instr->outputs[0]->id;
+      ReLinkFillConstant(in2instr, constant_name, fill_constant_set.at(key));
+    }
+
+    CinnBuilder builder("fill_constant_folding_builder");
+    for (auto& var : program->GetInputs()) {
+      builder.CreateInput(var);
+    }
+    for (int i = 0; i < program->size(); i++) {
+      if (remove_instrs.end() != remove_instrs.find(&(*program)[i])) continue;
+      builder.AppendInstruction((*program)[i]);
+    }
+    *program = builder.Build();
+  }
+
+ private:
+  static InputToOpMap GetInputToOpMap(Program* program) {
+    // `in2instr` is used to represent the mapping of Input to Instruction.
+    InputToOpMap in2instr;
+
+    for (int i = 0; i < program->size(); ++i) {
+      auto& instr = (*program)[i];
+
+      for (const auto& in : instr->inputs) {
+        in2instr[in->id].insert(&instr);
+      }
+    }
+    return in2instr;
+  }
+
+  static void ReLinkFillConstant(const InputToOpMap& in2instr, const std::string& input_var_name, Variable* out_var) {
+    if (!in2instr.count(input_var_name)) {
+      LOG(WARNING) << "Var [" << input_var_name << "] not used by other op ! ";
+      return;
+    }
+
+    VLOG(4) << "Try replace the input Var [" << input_var_name << "] to [" << (*out_var)->id
+            << "], because the fill_constant will be folding.";
+
+    const auto& output_ops = in2instr.at(input_var_name);
+    for (auto op : output_ops) {
+      auto find_input = [&](const std::string& input_name) {
+        return std::find_if(
+            (*op)->inputs.begin(), (*op)->inputs.end(), [&](const Variable& var) { return var->id == input_name; });
+      };
+
+      auto it = find_input(input_var_name);
+      // Why Loop : To avoid the op's inputs are the same variable !
+      while (it != (*op)->inputs.end()) {
+        // erase previous fill_constant output var and replace to new fill_constant output var
+        auto next_it = (*op)->inputs.erase(it);
+        // keep the input place same, it's very important
+        (*op)->inputs.insert(next_it, *out_var);
+        // try find next var if there exist same input
+        it = find_input(input_var_name);
+      }
```

Comment:
```suggestion
      // Why Loop : To avoid the op's inputs are the same variable !
      for (auto it = find_input(input_var_name); it != (*op)->inputs.end(); it = find_input(input_var_name)) {
        // erase previous fill_constant output var and replace to new fill_constant output var
        auto next_it = (*op)->inputs.erase(it);
        // keep the input place same, it's very important
        (*op)->inputs.insert(next_it, *out_var);
      }
```
这样是等价的吧？

Reply:
Done